Trello 7 u39ec i7 update tasks after user mapping

marcosraimondi1 opened the pull request:
Assets sync with tasks on change (update of user mappings or fullSync)

Possible Optmizations:
- Save mappings on IntegrationAssets: allows to only send assets that where affected by a change on the mappings. Now we are sending all.

marcosraimondi1 committed:
Trello-7U39ecI7: Added assigneeIds to integrationAssets and added contributorId to task given user mapping. (Might need optimization, now updating every task on mapping update)

marcosraimondi1 committed:
Trello-7U39ecl7: renaming

rozagerardo commented on src/main/kotlin/com/angorasix/projects/management/integrations/application/IntegrationAssetService.kt:
```diff
@@ -49,6 +48,7 @@ class IntegrationAssetService(
         sourceSyncId: String,
         projectManagementId: String,
         requestingContributor: DetailedContributor,
+        mappings: SourceSyncMappings,
```
Creo que pasaria el mappings.user `Map<String,String>` en vez de esto, para evitar manipular clases particulares del `SourceSync` aca que esta asociado al `IntegrationAsset`.

Mas adelante tmb podriamos ver de pasar estas configuraciones a una clase de dominio "common", si vemos que hay mucha configuracion que no es totalmente particular del `SourceSync`, si no que bastante compartida y que afecta como se manejan los `IntegrationAsset`s

Igual tampoco estaria del todo mal, pq los integration assets estan bastante ligados al source sync de todas formas. (si usasemos una base de datos relacional con ORM podriamos acceder al SourceSync y las configs directametne tmb)

rozagerardo commented on src/main/kotlin/com/angorasix/projects/management/integrations/application/IntegrationAssetService.kt:
```diff
@@ -109,6 +108,28 @@ class IntegrationAssetService(
         return persistedAssets
     }
 
+    suspend fun syncAssetsToTasks(
```
Evitemos mencionar `Task` siempre que se pueda en este servicio; conceptualmente los microservicios tienen que ser independientes uno de otros. 

Entiendo que ahora tenemos cosas de Tasks (de la libreria de commons) aca para comunicarse con ese servicio, pero idealmente (y posiblemente en algun momento lo hagamos) deberiamos mover la logica de envio del mensaje fuera de la capa de servicio (similar a como tenemos la capa de `presentation` para los mensajes de HTTP y el `messaging` para el ingreso de los mensajes). En esa capa esta bien hablar de "TaskDto" pq tenemos que hacer un mensaje para esa API.

Pero aca, con eso se iria toda la logica de mapeo a DTOs, y esto quedaria limpio con solo logica de negocio propia de este servicio.




Incluso en la logica que corre dentro de este metodo no hay nada relacionado a `Tasks`.  Yo llamaria al metodo `publishSyncingTasks` o algo asi, pq eso es lo que importa para este servicio: Lo publicamos, e indicamos que esta como "syncing", esperando a que se sincronice o reciba un Ack

rozagerardo commented on src/main/kotlin/com/angorasix/projects/management/integrations/application/IntegrationAssetService.kt:
```diff
@@ -165,6 +173,24 @@ class IntegrationAssetService(
         }
     }
+
+    private fun mapAssetsToTasks(assets: List<IntegrationAsset>, mappings: SourceSyncMappings) : List<A6InfraTaskDto> {
```
Diria que este caso es un buen candidato para definirlo como un extension function, algo como:
```
 private fun List<IntegrationAsset>.mapToTasks(mappings: SourceSyncMappings): List<A6InfraTaskDto> =
this.map{
    it.toTask()  
  }
```

Con `it.toTask()` siendo otra extension function `IntegrationAsset.toTask(mappings: SourceSyncMappings)`.

O posiblemente mejor, crear solo esa ultima extension function, y evitar la primera llamando directamenete a `updatedAssets.map { it.toTask(mappings) }`

rozagerardo commented:
for this:
```
Possible Optmizations:

Save mappings on IntegrationAssets: allows to only send assets that where affected by a change on the mappings. Now we are sending all.
```
We can discuss it. At first, I'd say that would make it more difficult to sync, and conceptually, we would be saying "this Integration Asset is assigned to 'this' A6 user" when actually there is no direct correspondence between an A6 user and an Integration Asset (e.g. a Trello asset).

What I would do to optimize this (in the future), is to use a more advanced filter of the Integration Assets we want to retrieve; i.e. instead of using `assetsService.findForSourceSyncId`, we would use a filter based on the source assignees that were affected by the change (if the corresponding A6 user changed for that Trello user mapping, then we want to include it in the filter).
But that would require changing the patch logic, bc we need to know the data before and after the patch operation.

A middle step would be filtering out all the Assets with no assignees (in that case, we know for sure that it was not updated based on the mapping).

marcosraimondi1 committed:
Trello-7U39ecI7: refactored solution, fixed finding existing assets

marcosraimondi1 committed:
Trello-7U39ecl7: fixing lint

marcosraimondi1 committed:
Trello-7U39ecl7: fixing linting (max parameters on publishUpdatedAssets)

marcosraimondi1 committed:
Trello-7U39ecI7: Update tasks if assets are synced

rozagerardo committed:
issue with Spring Cloud Function bug: https://github.com/spring-cloud/spring-cloud-function/issues/1247

rozagerardo committed:
Fixed bug with correspondence message handling
Cleaned up Assets with no admin logic - relying on mandatory SourceSyncContext obtained from SourceSync class
